numpy/typing/_array_like.py

```diff
@@ -0,0 +1,27 @@
+import sys
+from typing import Any, overload, Sequence, TYPE_CHECKING, Union
+
+from numpy import ndarray
+from ._dtype_like import DtypeLike
+
+if sys.version_info >= (3, 8):
+    from typing import Protocol
+    HAVE_PROTOCOL = True
+else:
+    try:
+        from typing_extensions import Protocol
+    except ImportError:
+        HAVE_PROTOCOL = False
+    else:
+        HAVE_PROTOCOL = True
+
+if TYPE_CHECKING or HAVE_PROTOCOL:
+    class _SupportsArray(Protocol):
+        @overload
+        def __array__(self, __dtype: DtypeLike = ...) -> ndarray: ...
+        @overload
+        def __array__(self, dtype: DtypeLike = ...) -> ndarray: ...
+else:
+    _SupportsArray = Any
+
+ArrayLike = Union[bool, int, float, complex, _SupportsArray, Sequence]
```

Comment:
It would be nice to support buffer protocols, but Python's typing doesn't support that yet. In the meantime I would suggest adding `memoryview` and a comment referencing https://github.com/python/typing/issues/593

Reply:
Since `memoryview`s are `Sequence`s we do allow them currently, e.g.

```
import numpy as np

x = b'foobar'
v = memoryview(x)
np.array(v)
```

passes mypy. I will add the comment though.

Reply:
We should add a test explicitly for memoryviews though; I'll open an issue.

Reply:
I'd highly recommend commenting on this B.P.O issue asking for support for a Buffer protocol in typing, as that is the better place than the typing issue: https://bugs.python.org/issue27501